File: Android/app/src/main/java/com/example/groceryapplication/copilot/WordPieceTokenizer.kt
package com.example.groceryapplication.copilot

import android.content.Context

/**
 * BERT WordPiece tokenizer for `all-MiniLM-L6-v2` (bert-base-uncased vocabulary).
 *
 * This is the Kotlin twin of the iOS `WordPieceTokenizer`, and both have to reproduce
 * HuggingFace's `BertTokenizer` exactly. The stored product vectors were authored with the
 * Python tokenizer, so any divergence here shifts query vectors into a slightly different
 * region of the space and degrades ranking in a way that looks like "semantic search is
 * mediocre" rather than like a bug. [CopilotDiagnostics] asserts agreement against
 * known-good token ids for the demo queries.
 *
 * Pipeline, matching `BasicTokenizer` + `WordpieceTokenizer`:
 *   1. clean text (drop control chars, normalize whitespace)
 *   2. pad CJK codepoints so each becomes its own token
 *   3. lowercase and strip accents (NFD, drop combining marks)
 *   4. split on whitespace, then split punctuation into separate tokens
 *   5. greedy longest-match-first subword lookup with `##` continuations
 */
class WordPieceTokenizer private constructor(
    private val vocab: Map<String, Int>
) {
    private val unkId = vocab["[UNK]"] ?: 100
    private val clsId = vocab["[CLS]"] ?: 101
    private val sepId = vocab["[SEP]"] ?: 102
    val padId = vocab["[PAD]"] ?: 0

    val vocabularySize: Int get() = vocab.size

    companion object {
        /** Longest token in bert-base-uncased is 18 chars; bounding the greedy scan keeps
         *  tokenization linear in word length instead of quadratic. */
        private const val MAX_TOKEN_CHARS = 24

        /** HuggingFace drops words longer than this to `[UNK]` before subwording. */
        private const val MAX_CHARS_PER_WORD = 100

        private const val VOCAB_ASSET = "copilot/minilm-vocab.txt"

        @Volatile
        private var instance: WordPieceTokenizer? = null

        /** Loads and caches the tokenizer. The vocabulary is ~230 KB, so it is read once. */
        fun shared(context: Context): WordPieceTokenizer =
            instance ?: synchronized(this) {
                instance ?: load(context).also { instance = it }
            }

        private fun load(context: Context): WordPieceTokenizer {
            val table = HashMap<String, Int>(31_000)
            context.assets.open(VOCAB_ASSET).bufferedReader().useLines { lines ->
                // One token per line, and the line number IS the token id — so blank lines
                // must still consume an index.
                lines.forEachIndexed { index, raw ->
                    table[raw.removeSuffix("\r")] = index
                }
            }
            return WordPieceTokenizer(table)
        }
    }

    data class Encoded(val ids: IntArray, val mask: IntArray)

    /**
     * Encodes text to fixed-length `input_ids` / `attention_mask` arrays.
     *
     * Truncates to [maxLength] including the `[CLS]`/`[SEP]` pair, which is what
     * `truncation=True, padding="max_length"` does on the Python side.
     */
    fun encode(text: String, maxLength: Int): Encoded {
        val ids = ArrayList<Int>(maxLength)
        ids.add(clsId)
        val budget = maxLength - 2      // room for [CLS] and [SEP]

        outer@ for (word in basicTokenize(text)) {
            for (piece in wordPieces(word)) {
                if (ids.size - 1 >= budget) break@outer
                ids.add(piece)
            }
        }
        ids.add(sepId)

        val real = ids.size
        val outIds = IntArray(maxLength) { if (it < real) ids[it] else padId }
        val outMask = IntArray(maxLength) { if (it < real) 1 else 0 }
        return Encoded(outIds, outMask)
    }

    /** Token ids without padding — used by the diagnostics parity check. */
    fun tokenIds(text: String, maxLength: Int): List<Int> {
        val encoded = encode(text, maxLength)
        return encoded.ids.take(encoded.mask.sum())
    }

    // MARK: - Basic tokenization

    /** Whitespace + punctuation splitting, lowercasing and accent stripping. */
    fun basicTokenize(text: String): List<String> {
        val cleaned = StringBuilder(text.length)
        var index = 0
        while (index < text.length) {
            val cp = text.codePointAt(index)
            index += Character.charCount(cp)
            when {
                // Drop NUL and the replacement char outright.
                cp == 0 || cp == 0xFFFD -> Unit
                isWhitespace(cp) -> cleaned.append(' ')
                isControl(cp) -> Unit
                // Surround CJK codepoints with spaces so each becomes its own token.
                isCJK(cp) -> cleaned.append(' ').appendCodePoint(cp).append(' ')
                else -> cleaned.appendCodePoint(cp)
            }
        }

        val out = ArrayList<String>()
        for (chunk in cleaned.toString().split(' ')) {
            if (chunk.isEmpty()) continue
            // Lowercase then strip combining marks — the order matters for locales where
            // casing changes which marks are produced. Locale.ROOT so a Turkish device
            // does not turn "I" into a dotless ı and silently change tokenization.
            val lowered = chunk.lowercase(java.util.Locale.ROOT)
            val decomposed = java.text.Normalizer.normalize(lowered, java.text.Normalizer.Form.NFD)
            val stripped = StringBuilder(decomposed.length)
            var j = 0
            while (j < decomposed.length) {
                val cp = decomposed.codePointAt(j)
                j += Character.charCount(cp)
                if (!isCombiningMark(cp)) stripped.appendCodePoint(cp)
            }
            out.addAll(splitPunctuation(stripped.toString()))
        }
        return out
    }

    private fun splitPunctuation(text: String): List<String> {
        val out = ArrayList<String>()
        val current = StringBuilder()
        var i = 0
        while (i < text.length) {
            val cp = text.codePointAt(i)
            i += Character.charCount(cp)
            if (isPunctuation(cp)) {
                if (current.isNotEmpty()) {
                    out.add(current.toString())
                    current.setLength(0)
                }
                out.add(String(Character.toChars(cp)))
            } else {
                current.appendCodePoint(cp)
            }
        }
        if (current.isNotEmpty()) out.add(current.toString())
        return out
    }

    // MARK: - WordPiece

    /**
     * Greedy longest-match-first subwording. Emits `[UNK]` for the whole word if any
     * position cannot be matched, matching HuggingFace rather than emitting partial pieces.
     */
    fun wordPieces(word: String): List<Int> {
        if (word.isEmpty()) return emptyList()
        val chars = word.toCharArray()
        if (chars.size > MAX_CHARS_PER_WORD) return listOf(unkId)

        val pieces = ArrayList<Int>()
        var start = 0
        while (start < chars.size) {
            var end = minOf(chars.size, start + MAX_TOKEN_CHARS)
            var matched: Int? = null
            while (end > start) {
                val candidate = StringBuilder().apply {
                    if (start > 0) append("##")
                    append(chars, start, end - start)
                }.toString()
                val id = vocab[candidate]
                if (id != null) {
                    matched = id
                    break
                }
                end -= 1
            }
            val id = matched ?: return listOf(unkId)
            pieces.add(id)
            start = end
        }
        return pieces
    }

    // MARK: - Character classes (mirroring HuggingFace's helpers)

    private fun isWhitespace(cp: Int): Boolean {
        if (cp == ' '.code || cp == '\t'.code || cp == '\n'.code || cp == '\r'.code) return true
        return Character.getType(cp) == Character.SPACE_SEPARATOR.toInt()
    }

    private fun isControl(cp: Int): Boolean {
        if (cp == '\t'.code || cp == '\n'.code || cp == '\r'.code) return false
        return when (Character.getType(cp).toByte()) {
            Character.CONTROL, Character.FORMAT, Character.PRIVATE_USE,
            Character.SURROGATE, Character.UNASSIGNED -> true
            else -> false
        }
    }

    private fun isCombiningMark(cp: Int): Boolean = when (Character.getType(cp).toByte()) {
        Character.NON_SPACING_MARK, Character.COMBINING_SPACING_MARK,
        Character.ENCLOSING_MARK -> true
        else -> false
    }

    /**
     * HuggingFace treats all ASCII non-alphanumerics as punctuation, not just the Unicode
     * punctuation categories — so `$`, `+`, `^` and friends split too.
     */
    private fun isPunctuation(cp: Int): Boolean {
        if (cp in 33..47 || cp in 58..64 || cp in 91..96 || cp in 123..126) return true
        return when (Character.getType(cp).toByte()) {
            Character.CONNECTOR_PUNCTUATION, Character.DASH_PUNCTUATION,
            Character.START_PUNCTUATION, Character.END_PUNCTUATION,
            Character.INITIAL_QUOTE_PUNCTUATION, Character.FINAL_QUOTE_PUNCTUATION,
            Character.OTHER_PUNCTUATION -> true
            else -> false
        }
    }

    private fun isCJK(cp: Int): Boolean =
        cp in 0x4E00..0x9FFF || cp in 0x3400..0x4DBF || cp in 0x20000..0x2A6DF ||
        cp in 0x2A700..0x2B73F || cp in 0x2B740..0x2B81F || cp in 0x2B820..0x2CEAF ||
        cp in 0xF900..0xFAFF || cp in 0x2F800..0x2FA1F
}

File: iOS/GroceryApp/Copilot/WordPieceTokenizer.swift
import Foundation

/// BERT WordPiece tokenizer for `all-MiniLM-L6-v2` (bert-base-uncased vocabulary).
///
/// This has to reproduce HuggingFace's `BertTokenizer` behaviour exactly. The stored
/// product vectors were authored with the Python tokenizer, so any divergence here
/// silently shifts query vectors into a slightly different region of the space and
/// degrades ranking in a way that looks like "semantic search is mediocre" rather
/// than like a bug. `CopilotDiagnostics` asserts agreement against known-good token
/// ids for the demo queries.
///
/// Pipeline, matching `BasicTokenizer` + `WordpieceTokenizer`:
///   1. clean text (drop control chars, normalize whitespace)
///   2. pad CJK codepoints so each becomes its own token
///   3. lowercase and strip accents (NFD, drop combining marks)
///   4. split on whitespace, then split punctuation into separate tokens
///   5. greedy longest-match-first subword lookup with `##` continuations
final class WordPieceTokenizer {

    enum TokenizerError: Error, LocalizedError {
        case vocabularyMissing(String)

        var errorDescription: String? {
            switch self {
            case .vocabularyMissing(let name):
                return "Tokenizer vocabulary '\(name)' is missing from the app bundle."
            }
        }
    }

    private let vocab: [String: Int32]
    private let unkId: Int32
    private let clsId: Int32
    private let sepId: Int32
    let padId: Int32
    /// Longest token in bert-base-uncased is 18 chars; bounding the greedy scan keeps
    /// tokenization linear in word length instead of quadratic.
    private let maxTokenChars = 24
    /// HuggingFace drops words longer than this to `[UNK]` before subwording.
    private let maxCharsPerWord = 100

    init(vocabularyResource: String = "minilm-vocab", bundle: Bundle = .main) throws {
        guard let url = bundle.url(forResource: vocabularyResource, withExtension: "txt"),
              let raw = try? String(contentsOf: url, encoding: .utf8) else {
            throw TokenizerError.vocabularyMissing(vocabularyResource)
        }

        // One token per line, and the line number IS the token id — so blank lines
        // must still consume an index. Only a trailing newline is dropped.
        var lines = raw.split(separator: "\n", omittingEmptySubsequences: false)
        if lines.last?.isEmpty == true { lines.removeLast() }

        var table: [String: Int32] = [:]
        table.reserveCapacity(lines.count)
        for (index, line) in lines.enumerated() {
            let token = line.hasSuffix("\r") ? String(line.dropLast()) : String(line)
            table[token] = Int32(index)
        }

        self.vocab = table
        self.unkId = table["[UNK]"] ?? 100
        self.clsId = table["[CLS]"] ?? 101
        self.sepId = table["[SEP]"] ?? 102
        self.padId = table["[PAD]"] ?? 0
    }

    var vocabularySize: Int { vocab.count }

    /// Encodes text to fixed-length `input_ids` / `attention_mask` arrays.
    ///
    /// Truncates to `maxLength` including the `[CLS]`/`[SEP]` pair, which is what
    /// `truncation=True, padding="max_length"` does on the Python side.
    func encode(_ text: String, maxLength: Int) -> (ids: [Int32], mask: [Int32]) {
        var ids: [Int32] = [clsId]
        let budget = maxLength - 2      // room for [CLS] and [SEP]

        outer: for word in basicTokenize(text) {
            for piece in wordPieces(word) {
                if ids.count - 1 >= budget { break outer }
                ids.append(piece)
            }
        }
        ids.append(sepId)

        var mask = [Int32](repeating: 1, count: ids.count)
        if ids.count < maxLength {
            let padCount = maxLength - ids.count
            ids.append(contentsOf: [Int32](repeating: padId, count: padCount))
            mask.append(contentsOf: [Int32](repeating: 0, count: padCount))
        }
        return (ids, mask)
    }

    // MARK: - Basic tokenization

    /// Whitespace + punctuation splitting, lowercasing and accent stripping.
    func basicTokenize(_ text: String) -> [String] {
        var cleaned = String(String.UnicodeScalarView(text.unicodeScalars.compactMap { scalar in
            // Drop NUL and the replacement char outright; map every other control
            // character and separator to a plain space.
            if scalar.value == 0 || scalar.value == 0xFFFD { return nil }
            if isWhitespace(scalar) { return " " }
            if isControl(scalar) { return nil }
            return scalar
        }))

        // Surround CJK codepoints with spaces so each becomes its own token.
        if cleaned.unicodeScalars.contains(where: isCJK) {
            var padded = String.UnicodeScalarView()
            for scalar in cleaned.unicodeScalars {
                if isCJK(scalar) {
                    padded.append(" ")
                    padded.append(scalar)
                    padded.append(" ")
                } else {
                    padded.append(scalar)
                }
            }
            cleaned = String(padded)
        }

        var out: [String] = []
        for chunk in cleaned.split(separator: " ", omittingEmptySubsequences: true) {
            // Lowercase then strip combining marks — the order matters for locales
            // where casing changes which marks are produced.
            let lowered = chunk.lowercased()
            let stripped = String(String.UnicodeScalarView(
                lowered.decomposedStringWithCanonicalMapping.unicodeScalars.filter {
                    !isCombiningMark($0)
                }
            ))
            out.append(contentsOf: splitPunctuation(stripped))
        }
        return out
    }

    private func splitPunctuation(_ text: String) -> [String] {
        var out: [String] = []
        var current = String.UnicodeScalarView()
        for scalar in text.unicodeScalars {
            if isPunctuation(scalar) {
                if !current.isEmpty {
                    out.append(String(current))
                    current = String.UnicodeScalarView()
                }
                out.append(String(scalar))
            } else {
                current.append(scalar)
            }
        }
        if !current.isEmpty { out.append(String(current)) }
        return out
    }

    // MARK: - WordPiece

    /// Greedy longest-match-first subwording. Emits `[UNK]` for the whole word if any
    /// position cannot be matched, matching HuggingFace rather than emitting partial
    /// pieces.
    func wordPieces(_ word: String) -> [Int32] {
        let chars = Array(word)
        if chars.isEmpty { return [] }
        if chars.count > maxCharsPerWord { return [unkId] }

        var pieces: [Int32] = []
        var start = 0
        while start < chars.count {
            var end = min(chars.count, start + maxTokenChars)
            var matched: Int32?
            while end > start {
                var candidate = String(chars[start..<end])
                if start > 0 { candidate = "##" + candidate }
                if let id = vocab[candidate] {
                    matched = id
                    break
                }
                end -= 1
            }
            guard let id = matched else { return [unkId] }
            pieces.append(id)
            start = end
        }
        return pieces
    }

    // MARK: - Character classes (mirroring HuggingFace's helpers)

    private func isWhitespace(_ s: Unicode.Scalar) -> Bool {
        if s == " " || s == "\t" || s == "\n" || s == "\r" { return true }
        return s.properties.generalCategory == .spaceSeparator
    }

    private func isControl(_ s: Unicode.Scalar) -> Bool {
        if s == "\t" || s == "\n" || s == "\r" { return false }
        switch s.properties.generalCategory {
        case .control, .format, .privateUse, .surrogate, .unassigned: return true
        default: return false
        }
    }

    private func isCombiningMark(_ s: Unicode.Scalar) -> Bool {
        switch s.properties.generalCategory {
        case .nonspacingMark, .spacingMark, .enclosingMark: return true
        default: return false
        }
    }

    /// HuggingFace treats all ASCII non-alphanumerics as punctuation, not just the
    /// Unicode punctuation categories — so `$`, `+`, `^` and friends split too.
    private func isPunctuation(_ s: Unicode.Scalar) -> Bool {
        let v = s.value
        if (v >= 33 && v <= 47) || (v >= 58 && v <= 64)
            || (v >= 91 && v <= 96) || (v >= 123 && v <= 126) { return true }
        switch s.properties.generalCategory {
        case .connectorPunctuation, .dashPunctuation, .openPunctuation,
             .closePunctuation, .initialPunctuation, .finalPunctuation,
             .otherPunctuation:
            return true
        default:
            return false
        }
    }

    private func isCJK(_ s: Unicode.Scalar) -> Bool {
        let v = s.value
        return (0x4E00...0x9FFF).contains(v) || (0x3400...0x4DBF).contains(v)
            || (0x20000...0x2A6DF).contains(v) || (0x2A700...0x2B73F).contains(v)
            || (0x2B740...0x2B81F).contains(v) || (0x2B820...0x2CEAF).contains(v)
            || (0xF900...0xFAFF).contains(v) || (0x2F800...0x2FA1F).contains(v)
    }
}
